### PR TITLE
Use net6 previews for building

### PR DESCRIPTION
**Pull request by vzarytovskii**

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -57,7 +57,7 @@ variables:
   - name: VisualStudioDropName
     value: Products/$(System.TeamProject)/$(Build.Repository.Name)/$(Build.SourceBranchName)/$(Build.BuildNumber)
   - name: DotNetSdkVersion
-    value: '5.0.100'
+    value: '6.0.100-preview.3.21202.5'
```
[the remark this answers is not in the document]
Comment (brettfo) [reply azure-pipelines.yml:60]: Good call.  According to [this](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/tool/dotnet-core-tool-installer?view=azure-devops#task-inputs) we can use:

``` yaml
- task: UseDotNet@2
  displayName: install SDK
  inputs:
    packageType: sdk
    useGlobalJson: true
    installationPath: $(Agent.ToolsDirectory)/dotnet
```

In these places:

- https://github.com/dotnet/fsharp/blob/187c8e069fd8ec73e675c2b53f3000f34d9c379e/azure-pipelines.yml#L369
- https://github.com/dotnet/fsharp/blob/187c8e069fd8ec73e675c2b53f3000f34d9c379e/azure-pipelines.yml#L392
- https://github.com/dotnet/fsharp/blob/187c8e069fd8ec73e675c2b53f3000f34d9c379e/azure-pipelines.yml#L415

And then delete that variable.

**Comment by vzarytovskii** [comment]
Okay, according to this issue https://github.com/microsoft/azure-pipelines-tasks/issues/11840, `DotNetCLIV2` gonna search for *all_ `global.json` files inside the working directory (the whole tree) and will try to install all SDKs at once, that's why it is trying to install fake one (from tests) and fails.

**Comment by vzarytovskii** [comment]
According to https://github.com/microsoft/azure-pipelines-tasks/issues/11840, `UseDotNet@2` will attempt to collect all the SDKs from all `global.json` files in the whole file tree starting from working directory.
Since we're using some `global.json` with wrong SDK version files in (manual) tests, it tries to install them as well and fails.
For now I will just add the version back to yaml file.

**Comment by vzarytovskii** [comment]
There's something fishy going on when runnning some of the cambridge tests:
```
  Failed array-FSI_BASIC [4 s]
  Error Message:
   Error running command 'C:\Users\_\code\fsharp\fsharp\tests\FSharp.Test.Utilities\..\..\.dotnet\dotnet.exe' with args 'build /t:RunFSharpScript' in directory 'C:\Users\_\code\fsharp\fsharp\artifacts\Temp\FSharp.Cambridge\av0yr0xc.s2g'.
   
...
   Microsoft (R) Build Engine version 16.10.0-preview-21181-07+073022eb4 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  Restored C:\Users\vzari\code\fsharp\fsharp\artifacts\Temp\FSharp.Cambridge\av0yr0xc.s2g\iz0jkuh5.2fe.fsproj (in 39 ms).
C:\Users\vzari\code\fsharp\fsharp\tests\fsharp\core\array\test.fsx(862,17): warning FS1204: This function is for use by compiled F# code and should not be used directly [C:\Users\vzari\code\fsharp\fsharp\artifacts\Temp\FSharp.Cambridge\av0yr0xc.s2g\iz0jkuh5.2fe.fsproj]
C:\Users\vzari\code\fsharp\fsharp\tests\fsharp\core\array\test.fsx(945,17): warning FS1204: This function is for use by compiled F# code and should not be used directly [C:\Users\vzari\code\fsharp\fsharp\artifacts\Temp\FSharp.Cambridge\av0yr0xc.s2g\iz0jkuh5.2fe.fsproj]
error FS0193 : internal error : Operation is not valid due to the current state of the object. [C:\Users\vzari\code\fsharp\fsharp\artifacts\Temp\FSharp.Cambridge\av0yr0xc.s2g\iz0jkuh5.2fe.fsproj]
```

**Comment by vzarytovskii** [comment]
Getting some weird issue when running some of the cambridge single-test suites;
```
"C:\Users\vzari\code\fsharp\fsharp\artifacts\Temp\FSharp.Cambridge\5a4qpb5o.stb\e5i33gro.2iy.fsproj" (RunFSharpScript target) (1:7) ->
                   (RunFSharpScript target) -> 
                     error FS0193 : internal error : Operation is not valid due to the current state of the object.
```

**Comment by brettfo** [comment]
@vzarytovskii What's your current working directory in each case?  I'm betting there's an upstream `global.json` or `Directory.Build.props/.targets` that's causing something different to happen.  If these are running under `%TEMP%`, [we copy in empty `Directory.Build.props/.targets` files](https://github.com/dotnet/fsharp/blob/7f22a41c003511b8b5028da48d0330843f93baaf/eng/Build.ps1#L299-L302) to prevent directory crawling, maybe we need to also copy in a default `global.json`?

**Comment by vzarytovskii** [comment]
@brettfo 
> @vzarytovskii What's your current working directory in each case? I'm betting there's an upstream `global.json` or `Directory.Build.props/.targets` that's causing something different to happen. If these are running under `%TEMP%`, [we copy in empty `Directory.Build.props/.targets` files](https://github.com/dotnet/fsharp/blob/7f22a41c003511b8b5028da48d0330843f93baaf/eng/Build.ps1#L299-L302) to prevent directory crawling, maybe we need to also copy in a default `global.json`?

It's always the created directory with fsproj (we do it for every test "suite"), here what's in there:
```
    Directory: C:\Users\vzari\code\fsharp\fsharp\artifacts\Temp\FSharp.Cambridge\5a4qpb5o.stb


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----         5/21/2021   5:43 PM                obj
-a----         5/21/2021   5:43 PM        1140479 buildoutput.txt
-a----         5/21/2021   5:43 PM            343 Directory.Build.props
-a----         5/21/2021   5:43 PM            274 Directory.Build.targets
-a----         5/21/2021   5:43 PM            149 Directory.Overrides.targets
-a----         5/21/2021   5:43 PM           4002 e5i33gro.2iy.fsproj
-a----         5/13/2021  12:31 PM          39397 test.fsx
-a----         5/21/2021   6:02 PM              2 test.ok
```
When I run the same command from the directory, all works as expected, I presume something is messed up with the environment of the "runner":
```
C:\Users\vzari\code\fsharp\fsharp\tests\FSharp.Test.Utilities\..\..\.dotnet\dotnet.exe build /t:RunFSharpScript
```
We generate project and prop files here https://github.com/dotnet/fsharp/blob/7f22a41c003511b8b5028da48d0330843f93baaf/tests/fsharp/single-test.fs#L66

**Comment by KevinRansom** [comment]
@vzarytovskii this is a bug in dotnet fsi ...

````
c:\kevinransom\fsharp>artifacts\bin\fsi\Release\net5.0\fsi.exe

Microsoft (R) F# Interactive version 11.3.1.0 for F# 5.0
Copyright (c) Microsoft Corporation. All Rights Reserved.

For help type #help;;

>     let arr = Array2D.create 3 4 0
-     let _ = Array2D.get arr 0 0
- ;;


error FS0193: internal error: Operation is not valid due to the current state of the object.
>
````

**Comment by KevinRansom** [comment]
Bizarrely this no longer repros on my machine.

**Comment by KevinRansom** [comment]
@vzarytovskii --- I believe this is a bug in the runtime, I hooked up with Jan to confirm that.

**Comment by vzarytovskii** [comment]
> @vzarytovskii --- I believe this is a bug in the runtime, I hooked up with Jan to confirm that.

I've seen the email, thanks a lot Kevin.

**Comment by KevinRansom** [comment]
runtime bug#  : https://github.com/dotnet/runtime/issues/53300

**Comment by KevinRansom** [comment]
runtime PR here: https://github.com/dotnet/runtime/pull/54656

**Comment by vzarytovskii** [comment]
We have some [NativeInterop tests failing](https://dev.azure.com/dnceng/public/_build/results?buildId=1261048&view=logs&j=22fea640-1099-5f32-ec5d-316ad83f359a&t=4607427e-2085-572d-4b25-975ef0413f8f&l=2645) with:
```log
Failed FSharp.Core.UnitTests.NativeInterop.NativePtr.Test<SByte>(n0: 0, nusual: 123, nextreme: -128, nhex11: 17) [1 ms]
  Error Message:
   System.InvalidProgramException : Common Language Runtime detected an invalid program.
```

I didn't look at it in detail yet, @Happypig375, you've added those, does the error ring a bell?
I will be back on Thursday, can look at it then in detail.

**Comment by vzarytovskii** [comment]
I can't reproduce tests failures locally, it seems :|

**Comment by vzarytovskii** [comment]
Interesting, the following test
```fsharp
[<Fact>]
member _.NullPtrTest() =
    let nullPtr = NativePtr.ofNativeInt<Int32> 0n
    Assert.Equal(NativePtr.nullPtr, nullPtr)
```
results in
```
   System.InvalidProgramException : Common Language Runtime detected an invalid program.
```

`NativePtr.ofNativeInt<Int32> 0n` should result in emitting `(# "" 0n: nativeptr<Int32> #)`
and
`NativePtr.nullPtr` - `(# "ldnull" : nativeptr<Int32> #)`

Gonna take a look what's the IL we emit.

**Comment by vzarytovskii** [comment]
Okay, `NativePtr.nullPtr` emits `ldnull` for non-reference values, we should just use 0n, fixed.

**Comment by vzarytovskii** [comment]
https://github.com/dotnet/runtime/pull/54656 gonna end up in preview7, so this is blocked for now from merging, gonna update and see if it's building and running tests fine.

**Comment by vzarytovskii** [comment]
After talking to @KevinRansom and @brettfo disabling failing test as agreed, going to re-enable after preview7 is released.

**Comment by vzarytovskii** [comment]
@KevinRansom @brettfo We can merge it now (fsi array tests are disabled, going to re-enable once preview7 is released).

File: tests/FSharp.Test.Utilities/CompilerAssert.fs [hunk 2]
```diff
@@ -337,7 +337,7 @@ type CompilerAssert private () =
         "tfm": "net5.0",
         "framework": {
             "name": "Microsoft.NETCore.App",
-            "version": "5.0.0"
+            "version": "6.0"
```
Comment (brettfo) [inline tests/FSharp.Test.Utilities/CompilerAssert.fs:340]: This should probably stay at `5` and we should _also_ install the 5.0 runtime at CI time.

File: tests/fsharp/Compiler/Language/SpanTests.fs [hunk 3]
```diff
@@ -70,51 +70,6 @@ test ()
         
         CompilerAssert.RunScript script []
 
-    [<Test>]
-    let Script_ReadOnlySpanForInDo() =
```
Comment (brettfo) [inline tests/fsharp/Compiler/Language/SpanTests.fs:74]: Commented out or skipped instead of removed entirely?

File: tests/fsharp/Compiler/Language/SpanTests.fs [hunk 4]
```diff
@@ -70,51 +70,6 @@ test ()
         
         CompilerAssert.RunScript script []
```
Comment (KevinRansom) [inline tests/fsharp/Compiler/Language/SpanTests.fs:72]: Why are these deleted?

File: tests/fsharp/tools/eval/test.fsx [hunk 5]
```diff
@@ -2285,7 +2285,7 @@ module Query =
             F ([srcTy],[src])
 
     let MakeQueryableTake = 
-        let F = MakeGenericStaticMethod <@ System.Linq.Queryable.Take @> 
+        let F = MakeGenericStaticMethod <@ System.Linq.Queryable.Take : _ * int -> _  @>
```
Comment (KevinRansom) [inline tests/fsharp/tools/eval/test.fsx:2288]: Why is this necessary now?

File: eng/Build.ps1 [hunk 6]
```diff
@@ -483,7 +485,7 @@ try {
         TestUsingNUnit -testProject "$RepoRoot\tests\FSharp.Compiler.Service.Tests\FSharp.Compiler.Service.Tests.fsproj" -targetFramework $coreclrTargetFramework -testadapterpath "$ArtifactsDir\bin\FSharp.Compiler.Service.Tests\"
         TestUsingXUnit -testProject "$RepoRoot\tests\FSharp.Compiler.Private.Scripting.UnitTests\FSharp.Compiler.Private.Scripting.UnitTests.fsproj" -targetFramework $coreclrTargetFramework -testadapterpath "$ArtifactsDir\bin\FSharp.Compiler.Private.Scripting.UnitTests\"
         TestUsingNUnit -testProject "$RepoRoot\tests\FSharp.Build.UnitTests\FSharp.Build.UnitTests.fsproj" -targetFramework $coreclrTargetFramework -testadapterpath "$ArtifactsDir\bin\FSharp.Build.UnitTests\"
-        TestUsingXUnit -testProject "$RepoRoot\tests\FSharp.Core.UnitTests\FSharp.Core.UnitTests.fsproj" -targetFramework $coreclrTargetFramework -testadapterpath "$ArtifactsDir\bin\FSharp.Core.UnitTests\"
+        TestUsingXUnit -testProject "$RepoRoot\tests\FSharp.Core.UnitTests\FSharp.Core.UnitTests.fsproj" -targetFramework $coreclrTargetFramework -testadapterpath "$ArtifactsDir\bin\FSharp.Core.UnitTests\"
```
Comment (KevinRansom) [inline eng/Build.ps1:488]: trailing whitespace

**Review by KevinRansom — COMMENTED** [review]
Looks good, why the deleted span tests?

**Comment by vzarytovskii** [comment]
> Looks good, why the deleted span tests?

They got fixed in net6

File: tests/fsharp/tools/eval/test.fsx (hunk 5 above)
Comment (vzarytovskii) [reply tests/fsharp/tools/eval/test.fsx:2288]: There are two overloads now, this and the one accepting range, specifying type so it can be properly resolved.

File: tests/fsharp/Compiler/Language/SpanTests.fs (hunk 3 above)
Comment (vzarytovskii) [reply tests/fsharp/Compiler/Language/SpanTests.fs:74]: These were testing a regression, which got fixed with net6.

On a second thought I probably should just change the assert.

File: tests/fsharp/Compiler/Language/SpanTests.fs (hunk 4 above)
Comment (vzarytovskii) [reply tests/fsharp/Compiler/Language/SpanTests.fs:72]: These were testing a regression, which got fixed with net6.

On a second thought I probably should just change the assert.

**Comment by vzarytovskii** [comment]
> Looks good, why the deleted span tests?

These were testing a regression, which got fixed with net6.

On a second thought I probably should just change the assert.

**Comment by auduchinok** [comment]
@vzarytovskii What is the reason to require the preview for building FSharp.Compiler.Service?

**Comment by vzarytovskii** [comment]
> @vzarytovskii What is the reason to require the SDK preview for building FSharp.Compiler.Service.sln? It makes things much worse for external contributors.

We use sdk Arcade uses by default, which will allow us to test latest runtime/sdk faster and see if there were any regressions (we were hitting some earlier this year).

It is a good point though, I'm not sure how can we support more than one SDK.

**Comment by auduchinok** [comment]
> It is a fair point though, I'm not sure how can we support more than one SDK.

We can move the FCS solution back to a separate folder and have a separate global.json there.

**Comment by vzarytovskii** [comment]
> > It is a fair point though, I'm not sure how can we support more than one SDK.
> 
> We can move the FCS solution back to a separate folder and have a separate global.json there.

Yeah, I guess that would work, let quickly make the change. 

What SDK do you think should be used for the service?

**Comment by auduchinok** [comment]
> What SDK do you think should be used for the service?

I'd say the oldest suitable for building the solution (including on the CI) or the previously used, 5.0.300.

**Comment by vzarytovskii** [comment]
> > What SDK do you think should be used for the service?
> 
> I'd say the oldest suitable for building the solution (including on the CI) or the previously used one, 5.0.300.

Huh, I don't think we build service solution in CI (at least, not directly via build scripts, or in yml), probably should add those too.

**Comment by vzarytovskii** [comment]
Created https://github.com/dotnet/fsharp/issues/11933 for tracking it.

**Comment by auduchinok** [comment]
> probably should add those too.

That would be really great if it's indeed not being built currently! 🙂
